extensions/pkg/controller/worker/genericactuator/machine_controller_manager.go

```diff
@@ -124,8 +125,58 @@ func (a *genericActuator) waitUntilMachineControllerManagerIsDeleted(ctx context
 }
 
 func (a *genericActuator) scaleMachineControllerManager(ctx context.Context, logger logr.Logger, worker *extensionsv1alpha1.Worker, replicas int32) error {
+	deployment := &appsv1.Deployment{}
+	if err := a.client.Get(ctx, kutil.Key(worker.Namespace, McmDeploymentName), deployment); err != nil {
+		if apierrors.IsNotFound(err) {
+			logger.Info("Machine-controller-manager is not deployed yet")
+			return nil
+		}
+		return err
+	}
 	logger.Info("Scaling machine-controller-manager", "replicas", replicas)
-	return kubernetes.ScaleDeployment(ctx, a.client, kutil.Key(worker.Namespace, McmDeploymentName), replicas)
+	if err := kubernetes.ScaleDeployment(ctx, a.client, kutil.Key(worker.Namespace, McmDeploymentName), replicas); err != nil {
+		return err
+	}
+	return a.waitUntilMachineControllerManagerIsScaled(ctx, worker)
+}
+
+func (a *genericActuator) waitUntilMachineControllerManagerIsScaled(ctx context.Context, worker *extensionsv1alpha1.Worker) error {
+	deployment := &appsv1.Deployment{}
+
+	if err := retry.UntilTimeout(ctx, 5*time.Second, 300*time.Second, func(ctx context.Context) (done bool, err error) {
+		if err := a.client.Get(ctx, kutil.Key(worker.Namespace, McmDeploymentName), deployment); err != nil {
+			return retry.SevereError(err)
+		}
+		if deployment.Generation != deployment.Status.ObservedGeneration {
+			return retry.MinorError(fmt.Errorf("machine-controller-manager not observed at latest generation (%d/%d)",
+				deployment.Status.ObservedGeneration, deployment.Generation))
+		}
+
+		replicas := int32(0)
+		if deployment.Spec.Replicas != nil {
+			replicas = *deployment.Spec.Replicas
+		}
+		if replicas != deployment.Status.UpdatedReplicas {
+			return retry.MinorError(fmt.Errorf("machine-controller-manager does not have enough updated replicas (%d/%d)",
+				deployment.Status.UpdatedReplicas, replicas))
+		}
+		if replicas != deployment.Status.Replicas {
+			return retry.MinorError(fmt.Errorf("machine-controller-manager deployment has outdated replicas"))
+		}
+		if replicas != deployment.Status.AvailableReplicas {
+			return retry.MinorError(fmt.Errorf("machine-controller-manager does not have enough available replicas (%d/%d",
+				deployment.Status.AvailableReplicas, replicas))
+		}
+
+		return retry.Ok()
+	}); err != nil {
+		var retryError *retry.Error
+		if !errors.As(err, &retryError) {
+			return err
+		}
+	}
+
+	return nil
```

Comment:
Can you move https://github.com/gardener/gardener/blob/03794c4fab012f16625540cd2989ab8f328be93d/pkg/operation/botanist/waiter.go#L253-L275 to `pkg/utils/kubernetes` and then call it here?

Reply:
Thanks for pointing this out, it looks much better